cypress-io/cypress
# feat: [BREAKING] make cypress/react default to React 18 support

- Closes #24160

### User facing changelog


`cypress/react` now defaults to React 18. For React <= 17 support, use `cypress/react17`.

### Additional details


*Note*: this PR targets `release/12.0.0`.

As part as our [evergreen strategy](https://github.com/cypress-io/prod-eng-docs/blob/main/team-component-testing/ecosytemsupport.md#how-we-will-handle-updates) with Cy 11 we want `cypress/react` to work with the latest React version, which is now React 18. We should leave `cypress/react18` so that there are no dead ends, as well as add a `cypress/react17` for users who are on React 16 and 17 (like the Cypress monorepo).

### Steps to test


Note: you probably want to use the pre-release binaries. Grab them from the bot comment on the latest commit.

- Try making a new React project (CRA or Vite) that uses React 18. Install the binary - it should work out of the box. It will scaffold `cypress/react` in the `support/component.js` - not `cypress/react18`, like we currently do.
- Try making a new React project (CRA or Vite) that uses React 17. Install the binary - it should work out of the box. It will scaffold `cypress/react17` in the `support/component.js` - not `cypress/react`, like we currently do.
- Try an existing project using React 17 working with Cypress 11. Upgrade to Cypress 12 by installing the binary - it should NOT work out of the box. You get VERY confusing error. I do not have any way around this - `cypress/react` imports `react-dom/client`, which does NOT exist in React <= 17. Compilation fails. I am open to suggestions to improve this. For now, we need to have good messaging and point users to the migration guide. This will be discussed at the product sync meeting later this week.

Note: I highlighted the Vite error in the terminal - the launchpad error is basically impossible to decipher. 

![image](https://user-images.githubusercontent.com/19196536/204223887-21a00218-491c-48a0-bd90-000edcb41ff0.png)

I am not particularly happy with this. The alternatives are

- migration guide (text based). Cheap and easy, but not as discoverable.
- migration UI (expensive, time consuming to build)
- change our strategy around evergreen versioning 

### How has the user experience changed?


### PR Tasks


- [ TODO ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)?  *will do pending product review*
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
